src/box/memtx_tree.cc

```diff
@@ -684,56 +691,31 @@ tree_iterator_start_raw(struct iterator *iterator, struct tuple **ret)
 		 */
 		memtx_tree_iterator_prev(tree, &it->tree_iterator);
 	}
-
-	if (!equals && (type == ITER_EQ || type == ITER_REQ)) {
-		/*
-		 * Found nothing, iteration will be stopped now. That is the
-		 * last chance to record that the transaction have read the key.
-		 */
-		if (key_is_full) {
-			memtx_tx_track_point(txn, space, idx, it->key_data.key);
-			return 0;
-		}
-		/* it->tree_iterator is positioned on successor of a key! */
-		struct memtx_tree_data<USE_HINT> *res =
-			memtx_tree_iterator_get_elem(tree, &it->tree_iterator);
-		struct tuple *successor = res == NULL ? NULL : res->tuple;
-		memtx_tx_track_gap(txn, space, idx, successor, type,
-				   it->key_data.key, it->key_data.part_count);
-		return 0;
-	}
-
-	struct memtx_tree_data<USE_HINT> *res =
-		memtx_tree_iterator_get_elem(tree, &it->tree_iterator);
-	uint32_t mk_index = 0;
-	if (res != NULL) {
-		*ret = res->tuple;
+	res = memtx_tree_iterator_get_elem(tree, &it->tree_iterator);
+	bool empty_result = !equals && (type == ITER_EQ || type == ITER_REQ);
+	if (res != NULL && !empty_result) {
 		tree_iterator_set_current(it, res);
 		tree_iterator_set_next_method(it);
 		bool is_multikey = iterator->index->def->key_def->is_multikey;
-		mk_index = is_multikey ? (uint32_t)res->hint : 0;
+		uint32_t mk_index = is_multikey ? (uint32_t)res->hint : 0;
+		/*
+		 * We need to clarify the result tuple before story garbage
+		 * collection, otherwise it could get cleaned there.
+		 */
+		*ret = memtx_tx_tuple_clarify(txn, space, res->tuple, idx,
+					      mk_index);
 	}
-
-	if ((!key_is_full || (type != ITER_EQ && type != ITER_REQ)) &&
-	    memtx_tx_manager_use_mvcc_engine) {
-		/* it->tree_iterator is positioned on successor of a key! */
-		struct tuple *successor = res == NULL ? NULL : res->tuple;
-
-/********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND START*********/
-		memtx_tx_track_gap(in_txn(), space, idx, successor, type,
+	if (key_is_full && empty_result)
+		memtx_tx_track_point(txn, space, idx, it->key_data.key);
+	if (!key_is_full || ((type == ITER_GE || type == ITER_LE) && !equals) ||
+	    (type == ITER_GT || type == ITER_LT))
+		memtx_tx_track_gap(txn, space, idx, successor, type,
 				   it->key_data.key, it->key_data.part_count);
-/*********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND END**********/
```

Comment:
Why have you removed `MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND`? It seems to me it can be useful for further maintenance.

Reply:
Fixed: retained these comments.